### ISSUE TITLE
Randomly Write Between Pooled Drives

### ISSUE THREAD

**Issue by Jumblemuddle** [trapexit/mergerfs#654]
General description
===================
I'm trying to pool together multiple [Rclone Google drive mounts](https://rclone.org/drive/). They're all mounted using different users for quota reasons and are thus eventually consistent. They're located at `/mnt/rclone$i`.
I've created a MergerFS mount to randomly write between these: `mergerfs -o allow_other,use_ino,category.action=rand,category.create=rand,category.search=ff /mnt/rclone/\* /mnt/mergerfs`

I'm sure there's a better way to do this; any guidance is appreciated. :)
Ideally, when writing/editing files, all the other mounts should be ignored except a randomly chosen one (as they should be exactly the same), and when reading files, it should be read from a random mount without touching the others.

Expected behavior
=================
When writing a file (`dd if=/dev/urandom of=$(mktemp XXXXXX) bs=1M count=1000`), only the randomly chosen Rclone mount would see any activity.

Actual behavior
===============
An Rclone mount is indeed chosen at random, however all the other mounts receive constant lookup requests for the particular file. (This always results in `no such file or directory` obviously, since the file isn't done writing.)

I've also seen some strange behavior around removing files; it seems like the remove request is sent to all mounts even though `category.create=rand` is set.

System information
==================
```
mergerfs version: 2.25.1
FUSE library version: 2.9.8-mergerfs
fusermount version: 2.9.9
using FUSE kernel interface version 7.19
```
`rclone v1.46`
Mounts:
```
Filesystem           Size  Used Avail Use% Mounted on
crypt0009:/          1.0P     0  1.0P   0% /mnt/rclone/9
crypt0008:/          1.0P     0  1.0P   0% /mnt/rclone/8
crypt0007:/          1.0P     0  1.0P   0% /mnt/rclone/7
crypt0006:/          1.0P     0  1.0P   0% /mnt/rclone/6
crypt0003:/          1.0P     0  1.0P   0% /mnt/rclone/3
crypt0002:/          1.0P     0  1.0P   0% /mnt/rclone/2
crypt0005:/          1.0P     0  1.0P   0% /mnt/rclone/5
crypt0000:/          1.0P     0  1.0P   0% /mnt/rclone/0
crypt0004:/          1.0P     0  1.0P   0% /mnt/rclone/4
crypt0001:/          1.0P     0  1.0P   0% /mnt/rclone/1
0:1:2:3:4:5:6:7:8:9   10P     0   10P   0% /mnt/merged
```

**Comment by Jumblemuddle** [comment]
The reason I'm trying to do this instead of writing to a local folder and then moving the files regularly (using a random set of credentials) to the remote is that my local drive may not have space. I'd like to be able to utilize MergerFS's `moveonenospc` ability. When something tries to write a file, it should begin by writing the file locally through MergerFS to a local folder, but if the file ends up being too large, it should transparently start uploading directly to the remote.

Perhaps there's a better way to accomplish everything I'm trying to do here?
I want to split all remote traffic between multiple credentials (whether by multiple Rclone mounts or by a cron script which uploads using random Rclone credentials), and I want to theoretically handle files which are larger than my local drive can handle. Additionally, I'd like to have some sort of cache in front (I've been considering an Rclone cache backend in front of MergerFS for this).

**Comment by trapexit** [comment]
> An Rclone mount is indeed chosen at random, however all the other mounts receive constant lookup requests for the particular file. (This always results in no such file or directory obviously, since the file isn't done writing.)

I don't know what you mean. The mounts will receive (from mergerfs) any calls necessary to accomplish the particularly request. You'll need to provide more detail as to what exactly is being done.

> I've also seen some strange behavior around removing files; it seems like the remove request is sent to all mounts even though category.create=rand is set.

Sure. That's the default behavior for `unlink` and all `action` category functions. If you don't want that you have to change the policy.

There isn't a good way to do what you're attempting. There isn't a way to know the size of something before it's being written and that's kindof what you'd need. `moveonenospc` is generally not something I recommend because the larger the file the worse your experience. Especially with a remote cloud system as your target. It also only picks the drive with the most free space. It doesn't randomize them or anything. The point is to be a last ditch catch for path preservation.

rclone mount is not anything close to POSIX compatible. In the past it seemed like using it for writes would not result in the best behavior or performance. A better solution IMO would be to get a larger drive as a staging drive and use a script similar to that in the mergerfs docs to move the files as the drive fills or whatnot. Or maybe try using rclone cache and not store to local disk at all (though as their docs say that has some risks but so does rclone mount for writes in general).

**Comment by Jumblemuddle** [comment]
> I don't know what you mean. The mounts will receive (from mergerfs) any calls necessary to accomplish the particularly request. You'll need to provide more detail as to what exactly is being done.
I don't know much about the POSIX file standard, but the logs for every other Rclone mount was constantly logging `lookup` requests for the file that was being written.

> Sure. That's the default behavior for `unlink` and all `action` category functions. If you don't want that you have to change the policy.
Whoops, my bad; `category.action=rand` is set, but I'm still seeing these requests being sent to all 

> There isn't a good way to do what you're attempting. There isn't a way to know the size of something before it's being written and that's kindof what you'd need. `moveonenospc` is generally not something I recommend because the larger the file the worse your experience. Especially with a remote cloud system as your target. It also only picks the drive with the most free space. It doesn't randomize them or anything. The point is to be a last ditch catch for path preservation.
Hmm, good to know.

> rclone mount is not anything close to POSIX compatible. In the past it seemed like using it for writes would not result in the best behavior or performance. A better solution IMO would be to get a larger drive as a staging drive and use a script similar to that in the mergerfs docs to move the files as the drive fills or whatnot. Or maybe try using rclone cache and not store to local disk at all (though as their docs say that has some risks but so does rclone mount for writes in general).
I think that's the right call. I'll look into having a larger staging area and just accept that things might fail if I run out of disk space.

Is there any virtue to using MergerFS with multiple Rclone mounts for read only operations in a random manner? (Would MergerFS still look at all of them even if `category.search=rand` is set? I'd set `=RO` and only use the Rclone mounts for reading; all writing would be done when copying from the staging area to the remotes using Rclone directly.)

I guess a better alternative would be to try to patch Rclone to support randomly picking credentials. That way MergerFS would only be needed for the 'write to staging; read from Rclone' functionality.

**Comment by trapexit** [comment]
> Would MergerFS still look at all of them even if category.search=rand is set? 

What do you mean by "look at them"? It has to issue stat's to find out what exists. It can't just return some random location that doens't even have the content in question. What's the concern? Performance? I'd think rclone would have some caching. I could improve the behavior a bit by shuffling the branches and then checking if it exists rather than checking what exists then shuffling but that would only marginally change things.

> I guess a better alternative would be to try to patch Rclone to support randomly picking credentials.

You're still describing solutions rather than problems. What is the issue you're solving for? Bandwidth limitations?

**Comment by Jumblemuddle** [comment]
> What do you mean by "look at them"? It has to issue stat's to find out what exists. It can't just return some random location that doens't even have the content in question. What's the concern? Performance? I'd think rclone would have some caching. I could improve the behavior a bit by shuffling the branches and then checking if it exists rather than checking what exists then shuffling but that would only marginally change things.
The concern is excessive API calls being made by Rclone to Google Drive. I was under the impression that the shuffling was done prior to the existence checking. I don't think there's any need to change that behavior though, since it would only make sense for very specific use cases.

> You're still describing solutions rather than problems. What is the issue you're solving for? Bandwidth limitations?
My bad; the issue I'm solving for is API calls being made to Google Drive. They have a per-user quota which I believe can be bypassed by splitting the API calls between multiple users who all share the same folder.

I'm going to close this issue, as I don't believe there's any change that should be made to MergerFS.
If you think it makes sense to make the (existing -> shuffling) to (shuffling -> existing) change or add an options for that, feel free to open this back up or create a new issue.

**Comment by Jumblemuddle** [comment]
Admitedly there might still be an issue around `category.action=rand` making requests to all branches. I may be mistaken there though, and I'm planning on pivoting to another strategy anyways.

**Comment by trapexit** [comment]
> Admitedly there might still be an issue around category.action=rand making requests to all branches.

How so? It only queries each branch to find what to work on. If 2 of 4 drives have the file to act on and you don't query the drives then it would act on a drive without the file 50% of the time.

**Comment by trapexit** [comment]
> The concern is excessive API calls being made by Rclone to Google Drive.

Doesn't rclone cache and misc settings help manage that? mergerfs shouldn't be making unnecessary calls but it has to query to find what's there. There is misc caching abilities as defined in the docs but that only goes so far. rclone is the best place to manage that kind of stuff.

**Comment by Jumblemuddle** [comment]
I'm thinking about this all wrong.
MergerFS doesn't make sense for multiple drives which replicate amongst themselves. A single path wouldn't normally be in two branches at once, right? And if it is in both places, then a `rm $path` should definitely make the remove request to both branches.

Rclone does have caching settings, but for my use case of only wanting to use it to split calls between different routes to the same data, the caching has to be minimal. Otherwise MergerFS would get presumably get confused when it remove a file, and the file took several seconds to disappear from the other branches (though this was my mistaken understanding; the file would actually be removed from all branches resulting in unnecessary API calls to delete the same file). Also, when creating/changing a file, it would take several seconds for the change to appear on the other branches.

Basically it all boils down to me being dumb and trying to use MergerFS for something completely counter to what it was designed to do. :P
Thanks for being patient. :)

**Comment by trapexit** [comment]
> A single path wouldn't normally be in two branches at once, right?
Paths are regularly on multiple paths. Files? Less so. 

> And if it is in both places, then a rm $path should definitely make the remove request to both branches.

Only if `unlink` policy is set to all/epall (which it is by default). The policy is what dictates what gets acted on.

**Comment by Jumblemuddle** [comment]
> Paths are regularly on multiple paths. Files? Less so.

Right, under normal MergerFS usage, a file would never be written to multiple branches, right?

> Only if `unlink` policy is set to all/epall (which it is by default). The policy is what dictates what gets acted on.

Is `category.unlink` distinct from `category.all`? I assumed my setting `category.all=rand` would include that.

**Comment by trapexit** [comment]
> Right, under normal MergerFS usage, a file would never be written to multiple branches, right?

Correct. mergerfs does not write to multiple files at once (IE... file level RAID1).

>Is category.unlink distinct from category.all? I assumed my setting category.all=rand would include that.

Not sure what you're talking about. There are only 3 categories. `search`, `action`, and `create`.

**Comment by Jumblemuddle** [comment]
> Not sure what you're talking about. There are only 3 categories. `search`, `action`, and `create`.

That was my understanding; any idea why I would be seing remove requests to all branches when `category.all=rand` is set? I may have been misreading the Rclone logs, but it certainly seemed like the same request was being sent to all of them.

**Comment by trapexit** [comment]
It could be a bug. I will check. I doubt anyone ever changes the action policy to rand tbh.

But why do you want that behavior? You're sync'ing the delete out of band? If you have 4 branches and 4 files.... anything but `all` or `epall` will remove a single file of the 4. You'd still see the file in the pool till you unlinked it 4 times.

**Comment by Jumblemuddle** [comment]
It probably doesn't make sense for any reasonable usage.
I was originally surprised when a deletion appeared immediately on all mounts.
The original reasoning was that I only wanted one remote to make the delete API request and let that eventually propogate to the other mounts out of band.

**Comment by trapexit** [comment]
Ah. It's totally legit for it to work that way it's just a bit weird. I'll check out rand in a bit and see if there is an issue.

**Comment by trapexit** [comment]
Fix for the random issue in the new version coming out.
